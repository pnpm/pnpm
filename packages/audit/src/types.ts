export type AuditVulnerabilityCounts = {
  info: number,
  low: number,
  moderate: number,
  high: number,
  critical: number,
}

export type AuditResolution = {
  id: number,
  path: string,
  dev: boolean,
  optional: boolean,
  bundled: boolean,
}

export type AuditAction = {
  action: string,
  module: string,
  target: string,
  isMajor: boolean,
  resolves: Array<AuditResolution>,
}

export type AuditAdvisory = {
  findings: [
    {
      version: string,
      paths: Array<string>,
      dev: boolean,
      optional: boolean,
      bundled: boolean,
    },
  ],
  id: number,
  created: string,
  updated: string,
  deleted?: boolean,
  title: string,
  found_by: {
    name: string,
  },
  reported_by: {
    name: string,
  },
  module_name: string,
  cves: Array<string>,
  vulnerable_versions: string,
  patched_versions: string,
  overview: string,
  recommendation: string,
  references: string,
  access: string,
  severity: string,
  cwe: string,
  metadata: {
    module_type: string,
    exploitability: number,
    affected_components: string,
  },
  url: string,
}

export type AuditMetadata = {
  vulnerabilities: AuditVulnerabilityCounts,
  dependencies: number,
  devDependencies: number,
  optionalDependencies: number,
  totalDependencies: number,
}

export type AuditReport = {
  actions: Array<AuditAction>,
  advisories: {[id: string]: AuditAdvisory},
  muted: Array<Object>,
  metadata: AuditMetadata,
}

export type AuditActionRecommendation = {
  cmd: string,
  isBreaking: boolean,
  action: AuditAction,
}
