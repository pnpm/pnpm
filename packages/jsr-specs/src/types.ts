export interface JsrSpecBase {
  scope?: string
  name?: string
  pref?: string
}

/** Syntax: `jsr:<spec>` */
export interface JsrSpecWithoutAlias extends JsrSpecBase {
  scope?: undefined
  name?: undefined
  pref: string
}

/** Syntax: `jsr:@<scope>/<name>[@<spec>] */
export interface JsrSpecWithAlias extends JsrSpecBase {
  scope: string
  name: string
  pref?: string
}

export type JsrSpec = JsrSpecWithoutAlias | JsrSpecWithAlias

export interface ParsedJsrPackageName {
  scope: string
  name: string
}
