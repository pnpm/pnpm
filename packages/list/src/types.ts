import { DependenciesHierarchy } from 'dependencies-hierarchy'

export type PackageDependencyHierarchy = DependenciesHierarchy & {
  name?: string,
  version?: string,
  path: string,
}
