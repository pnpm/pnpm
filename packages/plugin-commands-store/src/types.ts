import { LogBase } from '@pnpm/logger'

export type ReporterFunction = (logObj: LogBase) => void
