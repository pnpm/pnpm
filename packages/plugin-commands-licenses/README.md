# @pnpm/plugin-commands-licenses

> The licenses command of pnpm

[![npm version](https://img.shields.io/npm/v/@pnpm/plugin-commands-licenses.svg)](https://www.npmjs.com/package/@pnpm/plugin-commands-licenses)

## Installation

```sh
pnpm add @pnpm/plugin-commands-licenses
```

## License

MIT
