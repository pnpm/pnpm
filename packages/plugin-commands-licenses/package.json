{
  "name": "@pnpm/plugin-commands-licenses",
  "version": "0.0.0",
  "description": "The licenses command of pnpm",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "files": [
    "lib",
    "!*.map"
  ],
  "engines": {
    "node": ">=14.6"
  },
  "scripts": {
    "lint": "eslint src/**/*.ts test/**/*.ts",
    "registry-mock": "registry-mock",
    "test:jest": "jest",
    "test:e2e": "registry-mock prepare && run-p -r registry-mock test:jest",
    "_test": "jest -u",
    "test": "pnpm run compile && pnpm run _test",
    "prepublishOnly": "pnpm run compile",
    "compile": "tsc --build && pnpm run lint --fix"
  },
  "repository": "https://github.com/pnpm/pnpm/blob/main/packages/plugin-commands-licenses",
  "keywords": [
    "pnpm7",
    "pnpm",
    "outdated"
  ],
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/pnpm/pnpm/issues"
  },
  "homepage": "https://github.com/pnpm/pnpm/blob/main/packages/plugin-commands-licenses#readme",
  "devDependencies": {
    "@pnpm/constants": "workspace:*",
    "@pnpm/filter-workspace-packages": "workspace:*",
    "@pnpm/plugin-commands-installation": "workspace:*",
    "@pnpm/plugin-commands-licenses": "workspace:*",
    "@pnpm/read-package-json": "workspace:*",
    "@pnpm/types": "workspace:*",
    "@pnpm/prepare": "workspace:*",
    "@pnpm/registry-mock": "3.1.0",
    "@types/ramda": "0.28.15",
    "@types/wrap-ansi": "^3.0.0",
    "@types/zkochan__table": "npm:@types/table@6.0.0"
  },
  "dependencies": {
    "@pnpm/cli-utils": "workspace:*",
    "@pnpm/licenses": "workspace:*",
    "chalk": "^4.1.2",
    "strip-ansi": "^6.0.1",
    "@zkochan/table": "^1.0.0",
    "ramda": "npm:@pnpm/ramda@0.28.1",
    "@pnpm/command": "workspace:*",
    "@pnpm/constants": "workspace:*",
    "@pnpm/lockfile-file": "workspace:*",
    "@pnpm/common-cli-options-help": "workspace:*",
    "@pnpm/config": "workspace:*",
    "@pnpm/error": "workspace:*",
    "render-help": "^1.0.2"
  },
  "funding": "https://opencollective.com/pnpm",
  "exports": {
    ".": "./lib/index.js"
  }
}
