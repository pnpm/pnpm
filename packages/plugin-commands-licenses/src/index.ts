import * as licenses from './licenses'

export { licenses }
