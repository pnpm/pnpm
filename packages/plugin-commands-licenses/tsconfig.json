{
  "extends": "@pnpm/tsconfig",
  "compilerOptions": {
    "outDir": "lib",
    "rootDir": "src"
  },
  "include": [
    "src/**/*.ts",
    "../../typings/**/*.d.ts"
  ],
  "references": [
    {
      "path": "../cli-utils"
    },
    {
      "path": "../command"
    },
    {
      "path": "../common-cli-options-help"
    },
    {
      "path": "../config"
    },
    {
      "path": "../constants"
    },
    {
      "path": "../error"
    },
    {
      "path": "../filter-workspace-packages"
    },
    {
      "path": "../license-scanner"
    },
    {
      "path": "../lockfile-file"
    },
    {
      "path": "../plugin-commands-installation"
    },
    {
      "path": "../read-package-json"
    },
    {
      "path": "../store-path"
    }
  ]
}
