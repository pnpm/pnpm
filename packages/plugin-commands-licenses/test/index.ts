/// <reference path="../../../typings/index.d.ts" />
import path from 'path'
import { licenses } from '@pnpm/plugin-commands-licenses'
import { REGISTRY_MOCK_PORT } from '@pnpm/registry-mock'
import stripAnsi from 'strip-ansi'
import { PackageManifest } from '@pnpm/types'

const REGISTRY_URL = `http://localhost:${REGISTRY_MOCK_PORT}`

const LICENSES_OPTIONS = {
  cacheDir: 'cache',
  fetchRetries: 1,
  fetchRetryFactor: 1,
  fetchRetryMaxtimeout: 60,
  fetchRetryMintimeout: 10,
  global: false,
  networkConcurrency: 16,
  offline: false,
  rawConfig: { registry: REGISTRY_URL },
  registries: { default: REGISTRY_URL },
  strictSsl: false,
  tag: 'latest',
  userAgent: '',
  userConfig: {},
}

jest.mock('@pnpm/read-package-json', () => ({
  readPackageJson: async (pkgPath: string) => {
    // mock the readPackageJson-call used in getPkgInfo to ensure
    // it returns a PackageManifest as in the tests we don't actually
    // have a content store or node_modules directory to fetch the
    // package.json files from
    return {
      license: 'MIT',
      homepage: 'https://pnpm.io',
      author: 'Jane Doe',
    } as PackageManifest
  },
}))

test('pnpm licenses', async () => {
  const { output, exitCode } = await licenses.handler({
    ...LICENSES_OPTIONS,
    dir: path.resolve('./test/fixtures/has-licenses'),
    long: false,
  })

  expect(exitCode).toBe(0)
  expect(stripAnsi(output)).toMatchSnapshot('show-packages')
})

test('pnpm licenses: show details', async () => {
  const { output, exitCode } = await licenses.handler({
    ...LICENSES_OPTIONS,
    dir: path.resolve('./test/fixtures/has-licenses'),
    long: true,
  })

  expect(exitCode).toBe(0)
  expect(stripAnsi(output)).toMatchSnapshot('show-packages-details')
})

test('pnpm licenses: output as json', async () => {
  const workspaceDir = path.resolve('./test/fixtures/has-licenses')
  const { output, exitCode } = await licenses.handler({
    ...LICENSES_OPTIONS,
    dir: workspaceDir,
    long: false,
    json: true,
  })

  expect(exitCode).toBe(0)
  const parsedOutput = JSON.parse(output)
  expect(Object.keys(parsedOutput)).toMatchSnapshot('found-license-types')
  const packagesWithMIT = parsedOutput['MIT']
  expect(packagesWithMIT.length).toBeGreaterThan(0)
  expect(Object.keys(packagesWithMIT[0])).toEqual(['name', 'version', 'path', 'license', 'vendorName', 'vendorUrl'])
  packagesWithMIT.forEach((pkg: { name: string, version: string, path: string }) => {
    const expectedPkgPath = path.join(workspaceDir, `${pkg.name}@${pkg.version}`.replace('/', '+'), 'node_modules', pkg.name)
    expect(pkg.path).toBe(expectedPkgPath)
  })
})

test('pnpm licenses: show details', async () => {
  await expect(
    licenses.handler({
      ...LICENSES_OPTIONS,
      dir: path.resolve('./test/fixtures/invalid'),
      long: true,
    })
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    '"No pnpm-lock.yaml found: Cannot check a project without a lockfile"'
  )
})
