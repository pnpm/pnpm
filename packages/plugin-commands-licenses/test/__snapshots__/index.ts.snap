// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pnpm licenses: output as json: found-license-types 1`] = `
[
  "MIT",
]
`;

exports[`pnpm licenses: show details: show-packages-details 1`] = `
"┌───────────────────────────────┬─────────┬─────────────────┐
│ Package                       │ License │ Details         │
├───────────────────────────────┼─────────┼─────────────────┤
│ @next/env                     │ MIT     │ Jane Doe        │
│                               │         │ https://pnpm.io │
├───────────────────────────────┼─────────┼─────────────────┤
│ @next/swc-android-arm-eabi    │ MIT     │ Jane Doe        │
│                               │         │ https://pnpm.io │
├───────────────────────────────┼─────────┼─────────────────┤
│ @next/swc-android-arm64       │ MIT     │ Jane Doe        │
│                               │         │ https://pnpm.io │
├───────────────────────────────┼─────────┼─────────────────┤
│ @next/swc-darwin-arm64        │ MIT     │ Jane Doe        │
│                               │         │ https://pnpm.io │
├───────────────────────────────┼─────────┼─────────────────┤
│ @next/swc-darwin-x64          │ MIT     │ Jane Doe        │
│                               │         │ https://pnpm.io │
├───────────────────────────────┼─────────┼─────────────────┤
│ @next/swc-freebsd-x64         │ MIT     │ Jane Doe        │
│                               │         │ https://pnpm.io │
├───────────────────────────────┼─────────┼─────────────────┤
│ @next/swc-linux-arm-gnueabihf │ MIT     │ Jane Doe        │
│                               │         │ https://pnpm.io │
├───────────────────────────────┼─────────┼─────────────────┤
│ @next/swc-linux-arm64-gnu     │ MIT     │ Jane Doe        │
│                               │         │ https://pnpm.io │
├───────────────────────────────┼─────────┼─────────────────┤
│ @next/swc-linux-arm64-musl    │ MIT     │ Jane Doe        │
│                               │         │ https://pnpm.io │
├───────────────────────────────┼─────────┼─────────────────┤
│ @next/swc-linux-x64-gnu       │ MIT     │ Jane Doe        │
│                               │         │ https://pnpm.io │
├───────────────────────────────┼─────────┼─────────────────┤
│ @next/swc-linux-x64-musl      │ MIT     │ Jane Doe        │
│                               │         │ https://pnpm.io │
├───────────────────────────────┼─────────┼─────────────────┤
│ @next/swc-win32-arm64-msvc    │ MIT     │ Jane Doe        │
│                               │         │ https://pnpm.io │
├───────────────────────────────┼─────────┼─────────────────┤
│ @next/swc-win32-ia32-msvc     │ MIT     │ Jane Doe        │
│                               │         │ https://pnpm.io │
├───────────────────────────────┼─────────┼─────────────────┤
│ @next/swc-win32-x64-msvc      │ MIT     │ Jane Doe        │
│                               │         │ https://pnpm.io │
├───────────────────────────────┼─────────┼─────────────────┤
│ @swc/helpers                  │ MIT     │ Jane Doe        │
│                               │         │ https://pnpm.io │
├───────────────────────────────┼─────────┼─────────────────┤
│ caniuse-lite                  │ MIT     │ Jane Doe        │
│                               │         │ https://pnpm.io │
├───────────────────────────────┼─────────┼─────────────────┤
│ client-only                   │ MIT     │ Jane Doe        │
│                               │         │ https://pnpm.io │
├───────────────────────────────┼─────────┼─────────────────┤
│ nanoid                        │ MIT     │ Jane Doe        │
│                               │         │ https://pnpm.io │
├───────────────────────────────┼─────────┼─────────────────┤
│ next                          │ MIT     │ Jane Doe        │
│                               │         │ https://pnpm.io │
├───────────────────────────────┼─────────┼─────────────────┤
│ picocolors                    │ MIT     │ Jane Doe        │
│                               │         │ https://pnpm.io │
├───────────────────────────────┼─────────┼─────────────────┤
│ postcss                       │ MIT     │ Jane Doe        │
│                               │         │ https://pnpm.io │
├───────────────────────────────┼─────────┼─────────────────┤
│ source-map-js                 │ MIT     │ Jane Doe        │
│                               │         │ https://pnpm.io │
├───────────────────────────────┼─────────┼─────────────────┤
│ styled-jsx                    │ MIT     │ Jane Doe        │
│                               │         │ https://pnpm.io │
├───────────────────────────────┼─────────┼─────────────────┤
│ tslib                         │ MIT     │ Jane Doe        │
│                               │         │ https://pnpm.io │
├───────────────────────────────┼─────────┼─────────────────┤
│ typescript (dev)              │ MIT     │ Jane Doe        │
│                               │         │ https://pnpm.io │
├───────────────────────────────┼─────────┼─────────────────┤
│ use-sync-external-store       │ MIT     │ Jane Doe        │
│                               │         │ https://pnpm.io │
└───────────────────────────────┴─────────┴─────────────────┘
"
`;

exports[`pnpm licenses: show-packages 1`] = `
"┌───────────────────────────────┬─────────┐
│ Package                       │ License │
├───────────────────────────────┼─────────┤
│ @next/env                     │ MIT     │
├───────────────────────────────┼─────────┤
│ @next/swc-android-arm-eabi    │ MIT     │
├───────────────────────────────┼─────────┤
│ @next/swc-android-arm64       │ MIT     │
├───────────────────────────────┼─────────┤
│ @next/swc-darwin-arm64        │ MIT     │
├───────────────────────────────┼─────────┤
│ @next/swc-darwin-x64          │ MIT     │
├───────────────────────────────┼─────────┤
│ @next/swc-freebsd-x64         │ MIT     │
├───────────────────────────────┼─────────┤
│ @next/swc-linux-arm-gnueabihf │ MIT     │
├───────────────────────────────┼─────────┤
│ @next/swc-linux-arm64-gnu     │ MIT     │
├───────────────────────────────┼─────────┤
│ @next/swc-linux-arm64-musl    │ MIT     │
├───────────────────────────────┼─────────┤
│ @next/swc-linux-x64-gnu       │ MIT     │
├───────────────────────────────┼─────────┤
│ @next/swc-linux-x64-musl      │ MIT     │
├───────────────────────────────┼─────────┤
│ @next/swc-win32-arm64-msvc    │ MIT     │
├───────────────────────────────┼─────────┤
│ @next/swc-win32-ia32-msvc     │ MIT     │
├───────────────────────────────┼─────────┤
│ @next/swc-win32-x64-msvc      │ MIT     │
├───────────────────────────────┼─────────┤
│ @swc/helpers                  │ MIT     │
├───────────────────────────────┼─────────┤
│ caniuse-lite                  │ MIT     │
├───────────────────────────────┼─────────┤
│ client-only                   │ MIT     │
├───────────────────────────────┼─────────┤
│ nanoid                        │ MIT     │
├───────────────────────────────┼─────────┤
│ next                          │ MIT     │
├───────────────────────────────┼─────────┤
│ picocolors                    │ MIT     │
├───────────────────────────────┼─────────┤
│ postcss                       │ MIT     │
├───────────────────────────────┼─────────┤
│ source-map-js                 │ MIT     │
├───────────────────────────────┼─────────┤
│ styled-jsx                    │ MIT     │
├───────────────────────────────┼─────────┤
│ tslib                         │ MIT     │
├───────────────────────────────┼─────────┤
│ typescript (dev)              │ MIT     │
├───────────────────────────────┼─────────┤
│ use-sync-external-store       │ MIT     │
└───────────────────────────────┴─────────┘
"
`;
